tellapart/aurproxy/source/sources/api.py

```diff
@@ -0,0 +1,336 @@
+
+  @property
+  def blueprint(self):
+    """Flask blueprint for HTTP API."""
+    return self._blueprint
+
+  def _build_blueprint(self, name):
+    """
+    Build a flask blueprint that exposes an HTTP API with endpoints for the
+    addition and removal of sources.
+
+    Args:
+        name - Required str name to be used in API routes.
+    """
+    blueprint = Blueprint(name, __name__)
+    api = Api(blueprint)
+    root = self
+
+    class DynamicSource(Resource):
+
+      def get(self, source_name):
+        managed_source = root._get_managed_source(source_name)
+        if not managed_source:
+          root._abort(source_name)
+        else:
+          return managed_source.configuration
+
+      def delete(self, source_name):
+        if not root._get_managed_source(source_name):
+          root._abort(source_name)
+        else:
+          root._delete_managed_source(source_name)
+        return '', 204
+
+      def put(self, source_name):
+        source_config = request.json.get('source')
+        expiration = self._parse_expiration(request.json.get('expiration',
+                                                             None))
+        if not root._get_managed_source(source_name):
+          root._add_managed_source(source_name, source_config, expiration)
+          response_code = 201
+        else:
+          root._delete_managed_source(source_name)
+          root._add_managed_source(source_name, source_config, expiration)
+          response_code = 200
+
+        return source_config, response_code
+
+      def _parse_expiration(self, expiration):
+        expiration_parsed = None
+        if expiration:
+          try:
+            expiration_parsed = dateutil.parser.parse(expiration)
+          except:
+            msg = 'Attempt to parse expiration failed!'
+            logger.exception(msg)
+            raise abort(400, message=msg)
+          if datetime.now() + timedelta(minutes=1) > expiration_parsed:
+            msg = 'Expiration must occur at least 1 minute in the future!'
+            logger.info(msg)
+            raise abort(400, message=msg)
+        return expiration_parsed
+
+    class DynamicSourceList(Resource):
+      def get(self):
+        return root._source_map.keys()
+
+    # List sources
+    api.add_resource(DynamicSourceList,
+                           '/api/source/{0}/sources'.format(name))
+
+    # Get detail for a source
+    api.add_resource(DynamicSource,
+                     '/api/source/{0}/source/<source_name>/'.format(name))
+
+    return blueprint
+
+  def _abort(self, name):
+    """
+    Abort an http request with a 404.
+
+    Args:
+        name - Required source name str to be used in error message.
+    """
+    message = "Source {0} doesn't exist".format(name)
+    abort(404, message=message)
+
+  @property
+  def endpoints(self):
+    """
+    List of SourceEndpoints for sources managed by this ApiSource instance.
+    """
+    eps = set()
```

Comment:
Might be easier/more readable to have a method that returns the sources instead of looking them up then doing an itertool.chain for the set initializer.